### PR TITLE
HTTPCLIENT-2080 - add getRetryInterval to HttpRequestRetryStrategy for use on retriable IOExceptions

### PR DESCRIPTION
**Pull request by ajbaldocchi**

### CODE REVIEW

**Comment by ok2c** [comment]
@ajbaldocchi I am not sure I understand the intent of this change. Why would one want a delay between re-execution of the request after an I/O error?

File: httpclient5/src/main/java/org/apache/hc/client5/http/HttpRequestRetryStrategy.java [hunk 1]
```diff
@@ -75,6 +75,22 @@
...
+     *
+     * @param request the request failed due to an I/O exception
+     * @param exception the exception that occurred
+     * @param execCount the number of times this method has been
+     *                  unsuccessfully executed
+     * @param context the context for the request execution
+     *
+     * @return the retry interval between subsequent retries
+     */
+    default TimeValue getRetryInterval(HttpRequest request, IOException exception, int execCount, HttpContext context) {
```
Comment (ok2c) [inline httpclient5/src/main/java/org/apache/hc/client5/http/HttpRequestRetryStrategy.java:90]: @ajbaldocchi Please the existing `getRetryInterval` method for default implementation of this method.

File: httpclient5/src/main/java/org/apache/hc/client5/http/impl/classic/HttpRequestRetryExec.java [hunk 2]
```diff
@@ -113,6 +113,24 @@ public ClassicHttpResponse execute(
                         LOG.info("Recoverable I/O exception ({}) caught when processing request to {}",
                                 ex.getClass().getName(), route);
                     }
+                    final TimeValue nextInterval = retryStrategy.getRetryInterval(request, ex, execCount, context);
+                    // Make sure the retry interval does not exceed the response timeout
+                    if (TimeValue.isPositive(nextInterval)) {
+                        final RequestConfig requestConfig = context.getRequestConfig();
+                        final Timeout responseTimeout = requestConfig.getResponseTimeout();
+                        if (responseTimeout != null && nextInterval.compareTo(responseTimeout) > 0) {
```
Comment (ok2c) [inline httpclient5/src/main/java/org/apache/hc/client5/http/impl/classic/HttpRequestRetryExec.java:121]: @ajbaldocchi the response timeout is not applicable here. Please drop this check.

**Review by ok2c — CHANGES_REQUESTED** [review]

File: httpclient5/src/main/java/org/apache/hc/client5/http/HttpRequestRetryStrategy.java (hunk 1 above)
Comment (ajbaldocchi) [reply httpclient5/src/main/java/org/apache/hc/client5/http/HttpRequestRetryStrategy.java:90]: Could you please clarify the change you would like here?

Comment (ok2c) [reply httpclient5/src/main/java/org/apache/hc/client5/http/HttpRequestRetryStrategy.java:90]: @ajbaldocchi My aplogies. Please disregard my comment. I overlooked that the new method takes `HttpRequest` as a parameter.

**Review by ok2c — APPROVED** [review]

**Comment by ok2c** [comment]
@michael-o Any objections to merging this change-set?

**Comment by michael-o** [comment]
Let me paraphrase your PR: Not only requests on HTTP level shall be repeatable, but you also want on `IOException` level decide how to to wait for the retry. Correct?

**Comment by ajbaldocchi** [comment]
Correct. For a retriable IOException the option to control how long to wait to initiate the retry is relevant to my use case.

**Pull request merged**